text/0000-floating-point-additional-precision.md

```diff
@@ -0,0 +1,141 @@
+states that:
+> The IEEE 754-2008 "binary32" and "binary64" floating-point types are f32 and f64, respectively.
+
+This RFC proposes updating that definition as follows:
+
+The `f32` and `f64` types represent the IEEE 754-2008 "binary32" and "binary64"
+floating-point types. Operations on those types must provide no less
+precision than the IEEE standard requires; such operations may provide *more*
+precision than the standard requires, such as by doing a series of operations
+with higher precision before storing a value of the desired precision.
+
+rustc may provide a codegen (`-C`) option to disable this behavior, such as `-C
+disable-extra-fp-precision`. Rust may also provide an attribute to disable this
+behavior from within code, such as `#[disable_extra_fp_precision]`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If Rust already provided bit-for-bit identical floating-point computations
+across platforms, this change could potentially allow floating-point
+computations to differ by platform (though never below the standards-required
+accuracy). However, standards-compliant implementations of math functions on
+floating-point values may already vary slightly by platform, sufficiently so to
+produce different binary results. This proposal can never make results *less*
+accurate, it can only make results *more* accurate.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could provide a separate set of types and allow extra accuracy in their
+operations; however, this would create ABI differences between floating-point
+functions, and the longer, less-well-known types seem unlikely to see
+widespread use.
+
+We could provide an option to enable extra accuracy for the default
+floating-point types, but disable it by default. This would leave the majority
+of floating-point code unable to use these optimizations, however; defaults
+matter, and the majority of code seems likely to use the defaults.
+
+We could do nothing, and require code to use `a.mul_add(b, c)` for
+optimization; however, this would not allow for similar future optimizations,
+and would not allow code to easily enable this optimization without substantial
+code changes.
+
+We could narrow the scope of optimization opportunities to *only* include
+floating-point contraction but not any other precision-increasing operations.
+See the [future possibilities](#future-possibilities) section for further
+discussion on this point.
+
+# Prior art
+[prior-art]: #prior-art
+
+This has precedent in several other languages and compilers:
+
+- [C11](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf) allows
+  this with the `STDC FP_CONTRACT` pragma enabled, and the default state
+  of that pragma is implementation-defined. GCC enables this pragma by
+  default, [as does the Microsoft C
+  compiler](https://docs.microsoft.com/en-us/cpp/preprocessor/fp-contract?view=vs-2019).
```

Comment:
What does clang do here?

Reply:
ICC generates FMA by default, just as GCC does.

I don't have access to IBM XLC. Based on scattered third-party reports, it sounds like its powerpc64 target generates `fmadd` instructions by default for `a * b + c`.

clang has the opposite default.

Reply:
I think this information should be mentioned in the RFC, otherwise it feels like all compilers agree by default on this, but that's not the case.